packages/kbn-migrations/README.md

```diff
@@ -0,0 +1,322 @@
+        { "remove" : { "index" : ".kibana-7.0.0-alpha1-e8c454203d8d4dbbc253fc5e28ae7ae36ec3c1a8", "alias" : ".kibana" } }
+    ]
+}
+'
+```
+
+Put the previous index in the "add" action and the current index in the "remove" action.
+
+Next, we need to ensure the previous index's migration status is 'migrated':
+
+```
+curl -X POST "localhost:9200/.kibana/doc/migration:migration-state/_update" -H 'Content-Type: application/json' -d'
+{
+  "script" : "ctx._source.migration.status = \"migrated\""
+}
+'
+```
+
+Finally, let's ensure that the Kibana index is writeable.
+
+```
+curl -X PUT "localhost:9200/.kibana/_settings" -H 'Content-Type: application/json' -d'
+{
+    "index" : {
+        "blocks.read_only": "false"
+    }
+}
+'
+```
+
+It should now be possible to start the previous version of Kibana.
+
+
+
+## Design choices and constraints
+
+- Not intended to be used to migrate large indices, as it pulls documents from Elasticsearch into Node and then back to Elasticsearch
+- The index is assumed to only have documents which conform to the saved-object-client conventions
+- Does not attempt to recover from failure of an individual migration
+  - If an individual migration (transform or seed) fails, the migration fails
+- Plugin ids are assumed to be stable. If a plugin's id changes, the migration system will:
+  - Will think that it has been removed and a brand new plugin has been added
+  - Will attempt to run all migrations for the "new" plugin
+- Migration order is assumed to be stable within a plugin
+  - So, the system will error in a scenario like this:
+    - A plugin defines migrations 'a', 'b', and 'c'
+    - Those migrations have been applied to an index
+    - The plugin is upgraded to a later version, which defines migrations 'a', 'c', 'b'
+    - In this case, the plugin's migrations have been reordered
+- Mappings are never removed
+  - This is because we need to support old indices which might have mappings that are no longer associated with a plugin
+  - We could put in logic to clean up mappings at the end of a migration:
+    - During migration, track what mappings are required by existing docs
+    - After migration, remove any mappings that are not defined by active plugins and that are not required by existing docs
+
+
+## Kibana-index specific notes
+
+Apart from the core migration logic in the `kbn-migrations` package, there is logic which is specific to the Kibana index itself.
+
+When Kibana starts, it will check to see if it needs to migrate its index. If it does, it will perform the migration prior to serving any HTTP requests. Only one instance of Kibana will perform migrations at any given time. All other Kibana instances will fail to start and will log an error indicating that the Kibana index is migrating. Once the Kibana index migration has completed, the other instances will be able to start successfully.
```

Comment:
Still yet to verify, but want to verify that we do not fail the other instances - we simply await a migrated state.

Reply:
You are right, and this readme is wrong. That changed after I wrote this, I think. Good catch (as are all the others)!